src/compiler/checker.ts

```diff
@@ -13301,9 +13332,31 @@ namespace ts {
                         neverType;
                 }
             }
+            if (!origin && includes & TypeFlags.Union) {
+                const namedUnions: Type[] = [];
+                addNamedUnions(namedUnions, types);
+                const reducedTypes: Type[] = [];
+                for (const t of typeSet) {
+                    if (!some(namedUnions, union => containsType((<UnionType>union).types, t))) {
+                        reducedTypes.push(t);
+                    }
+                }
+                if (!aliasSymbol && namedUnions.length === 1 && reducedTypes.length === 0) {
+                    return namedUnions[0];
+                }
+                // We create a denormalized origin type only when the union was created from one or more named unions
+                // (unions with alias symbols or origins) and when there is no overlap between those named unions.
+                const namedTypesCount = reduceLeft(namedUnions, (sum, union) => sum + (<UnionType>union).types.length, 0);

```

Comment:
Given:
```ts
type Foo = "a" | "b";
type Bar = "b" | "c";
declare let x: Foo | Bar;
```
Does this mean `x` will have type `"a" | "b" | "c"` because there is overlap? I'm not opposed (since its no worse than what we see today), but is there any other reason to disallow overlap?

Reply:
Yes, that's what it means. Without this check the denormalized form may actually be *more complex* than the normalized form, which is pointless. For example `AB | AC | AD | BC | BD | CD` vs. `A | B | C | D`. By ensuring there's no overlap, we know we're always displaying something that has fewer constituents (and thus presumably is simpler).